src/main/java/hudson/plugins/gradle/injection/MavenInjectionAware.java

```diff
@@ -0,0 +1,96 @@
+    // Maven system properties passed on the CLI to a Maven build
+    SystemProperty.Key GRADLE_ENTERPRISE_URL_PROPERTY_KEY = SystemProperty.Key.required("gradle.enterprise.url");
+    SystemProperty.Key BUILD_SCAN_UPLOAD_IN_BACKGROUND_PROPERTY_KEY = SystemProperty.Key.required("gradle.scan.uploadInBackground");
+    SystemProperty.Key MAVEN_EXT_CLASS_PATH_PROPERTY_KEY = SystemProperty.Key.required("maven.ext.class.path");
+    SystemProperty.Key GRADLE_ENTERPRISE_ALLOW_UNTRUSTED_SERVER_PROPERTY_KEY = SystemProperty.Key.optional("gradle.enterprise.allowUntrustedServer");
+
+    MavenOptsHandler MAVEN_OPTS_HANDLER = new MavenOptsHandler(
+            MAVEN_EXT_CLASS_PATH_PROPERTY_KEY,
+            BUILD_SCAN_UPLOAD_IN_BACKGROUND_PROPERTY_KEY,
+            GRADLE_ENTERPRISE_ALLOW_UNTRUSTED_SERVER_PROPERTY_KEY,
+            GRADLE_ENTERPRISE_URL_PROPERTY_KEY
+    );
+
+    default boolean isInjectionDisabledGlobally(InjectionConfig config) {
+        if (config.isDisabled() || !config.isInjectMavenExtension()) {
+            return true;
+        }
+
+        return InjectionUtil.isInvalid(InjectionConfig.checkRequiredUrl(config.getServer()));
+    }
+
+    default boolean isInjectionEnabledForNode(Node node) {
+        InjectionConfig config = InjectionConfig.get();
+
+        if (isInjectionDisabledGlobally(config)) {
+            return false;
+        }
+
+        Set<String> disabledNodes =
+                CollectionUtil.safeStream(config.getMavenInjectionDisabledNodes())
+                        .map(NodeLabelItem::getLabel)
+                        .filter(Objects::nonNull)
+                        .collect(Collectors.toSet());
+
+        Set<String> enabledNodes =
+                CollectionUtil.safeStream(config.getMavenInjectionEnabledNodes())
+                        .map(NodeLabelItem::getLabel)
+                        .filter(Objects::nonNull)
+                        .collect(Collectors.toSet());
+
+        return InjectionUtil.isInjectionEnabledForNode(node::getAssignedLabels, disabledNodes, enabledNodes);
+    }
+
+    default void setPreparedMavenProperties(Run<?, ?> run, TaskListener listener) {
+        try {
+            EnvVars environment = run.getEnvironment(listener);
+
+            String preparedMavenOpts = environment.get(JENKINSGRADLEPLUGIN_MAVEN_OPTS_PREPARED);
+            String preparedMavenPluginConfigExtClasspath = environment.get(JENKINSGRADLEPLUGIN_MAVEN_PLUGIN_CONFIG_EXT_CLASSPATH_PREPARED);
+            if (preparedMavenOpts != null && preparedMavenPluginConfigExtClasspath != null) {
+                run.addAction(new NodePreparedMavenEnvsAction(preparedMavenOpts, preparedMavenPluginConfigExtClasspath));
+            }
+        } catch (IOException | InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Action that holds Maven environment variables to be set in EnvironmentContributor
+     */
+    class NodePreparedMavenEnvsAction extends InvisibleAction {
```

Comment:
Maybe call it `MavenInjectionEnvVarsAction` to make the name more aligned with the env contributor?

Reply:
How about `MavenInjectionNodeEnvVarsAction`? Or do you think "Node" is redundant then?

Reply:
I think it's redundant. We are just transferring environment variables needed for the Maven auto-injection. It makes this data to be quite generic and not bound to any node.

Reply:
Yes, it's correct. It contains node-specific data, but the contributor (the receiver of the data) doesn't have any node-specific logic. It just gets the data (whatever this data is) and sets the required variables. I see this action as a generic envelope to just transfer data from the producer to the consumer. It's essentially the same as `EnvironmentContributingAction`, just tailored to transfer only specific variables (in form of fields).